### PR TITLE
ITE: drivers: gpio: it8xxx2: add support for 1.8v flag

### PR DESCRIPTION
**Pull request by Dino-Li**
This change make IT8xxx2 chips to support GPIO voltage
1.8v flag.

Signed-off-by: Dino Li <Dino.Li@ite.com.tw>

---
This change is [Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/36443)

### CODE REVIEW

**Review by brockus-zephyr — APPROVED** [review]

File: soc/riscv/riscv-ite/common/chip_chipregs.h [hunk 1]
```diff
@@ -1678,8 +1678,23 @@ struct wdt_it8xxx2_regs {
...
+#define IT8XXX2_GPIO_GCR20      ECREG(IT8XXX2_GPIO_BASE + 0xE5)
+#define IT8XXX2_GPIO_GCR21      ECREG(IT8XXX2_GPIO_BASE + 0xE6)
+#define IT8XXX2_GPIO_GCR22      ECREG(IT8XXX2_GPIO_BASE + 0xE7)
+#define IT8XXX2_GPIO_GCR23      ECREG(IT8XXX2_GPIO_BASE + 0xE8)
+#define IT8XXX2_GPIO_GCR24      ECREG(IT8XXX2_GPIO_BASE + 0xE9)
+#define IT8XXX2_GPIO_GCR30      ECREG(IT8XXX2_GPIO_BASE + 0xED)
+#define IT8XXX2_GPIO_GCR29      ECREG(IT8XXX2_GPIO_BASE + 0xEE)
+/* TODO: correct GRC1 to GCR1 */
 #define IT8XXX2_GPIO_GRC1       ECREG(IT8XXX2_GPIO_BASE + 0xF0)
-#define IT8XXX2_GPIO_GRC21      ECREG(IT8XXX2_GPIO_BASE + 0xE6)
```
Comment (brockus-zephyr) [inline soc/riscv/riscv-ite/common/chip_chipregs.h:1682]: Need to change the code that referred to this.

> /workdir/zephyr/soc/riscv/riscv-ite/it8xxx2/soc.c:36:2: error: 'IT8XXX2_GPIO_GRC21' undeclared (first use in this function); did you mean 'IT8XXX2_GPIO_GRC1'?
>    36 |  IT8XXX2_GPIO_GRC21 &= ~(BIT(0) | BIT(1));
>       |  ^~~~~~~~~~~~~~~~~~
>       |  IT8XXX2_GPIO_GRC1

Comment (brockus-zephyr) [reply soc/riscv/riscv-ite/common/chip_chipregs.h:1682]: Leaving my approved because this has to be fixed in order to get the required checks to complete.

File: drivers/gpio/gpio_ite_it8xxx2.c [hunk 2]
```diff
@@ -233,6 +349,8 @@ static int gpio_ite_configure(const struct device *dev,
 	volatile uint8_t *reg_gpdr = (uint8_t *)gpio_config->reg_gpdr;
 	volatile uint8_t *reg_gpcr = (uint8_t *)(gpio_config->reg_gpcr + pin);
 	volatile uint8_t *reg_gpotr = (uint8_t *)gpio_config->reg_gpotr;
+	volatile uint8_t *reg_1p8v = gpio_1p8v[gpio_config->idx][pin].reg_1p8v;
```
Comment (mnkp) [inline drivers/gpio/gpio_ite_it8xxx2.c:352]: Let's add an __ASSERT / __ASSERT_NO_MSG which verifies that `gpio_config->idx` and `pin` is withing limits.

File: dts/bindings/gpio/ite,it8xxx2-gpio.yaml [hunk 3]
```diff
@@ -8,16 +8,17 @@ compatible: "ite,it8xxx2-gpio"
 include: [gpio-controller.yaml, base.yaml]
 
 properties:
-    port-is-output:
-      type: boolean
-      description: Indicates if the port is an output port
-
```
Comment (mnkp) [inline dts/bindings/gpio/ite,it8xxx2-gpio.yaml:14]: Looks like unrelated change. Should be done in a separate commit.

File: drivers/gpio/gpio_ite_it8xxx2.c [hunk 4]
```diff
@@ -278,10 +396,14 @@ static int gpio_ite_configure(const struct device *dev,
...
 
-	/*
-	 * TODO: There are some gpios are 1.8v input at default.
-	 * Is there a configuration flag for 1.8/3.3v selection.
-	 */
+	/* 1.8V or 3.3V */
+	if (reg_1p8v != NULL) {
+		if ((flags & GPIO_VOLTAGE_MASK) == GPIO_VOLTAGE_1P8) {
+			*reg_1p8v |= mask_1p8v;
+		} else {
```
Comment (mnkp) [inline drivers/gpio/gpio_ite_it8xxx2.c:403]: Let's make sure only flags: GPIO_VOLTAGE_1P8, GPIO_VOLTAGE_3P3, GPIO_VOLTAGE_DEFAULT are accepted.

**Review by mnkp — COMMENTED** [review]
Let's change commit message to
```
drivers: gpio: it8xxx2: add support for GPIO_VOLTAGE_ flags

Support GPIO_VOLTAGE_1P8, GPIO_VOLTAGE_3P3 flags on IT8xxx2 chips.
```
And remove "Change-Id:" line.

**Review by jackrosenthal — APPROVED** [review]

File: dts/bindings/gpio/ite,it8xxx2-gpio.yaml [hunk 5]
```diff
@@ -8,16 +8,17 @@ compatible: "ite,it8xxx2-gpio"
...
     reg:
       required: true
 
     label:
       required: true
 
+    index:
```
Comment (galak) [inline dts/bindings/gpio/ite,it8xxx2-gpio.yaml:17]: Why do we need index, can't you compute this based on reg?

**Review by galak — DISMISSED** [review]
the dt index property doesn't seen required as seems like `0xf01601 - reg` will get you this.

**Review by Dino-Li — COMMENTED** [review]
> Let's change commit message to
> 
> ```
> drivers: gpio: it8xxx2: add support for GPIO_VOLTAGE_ flags
> 
> Support GPIO_VOLTAGE_1P8, GPIO_VOLTAGE_3P3 flags on IT8xxx2 chips.
> ```
> 
> And remove "Change-Id:" line.

Done.

File: drivers/gpio/gpio_ite_it8xxx2.c (hunk 4 above)
Comment (Dino-Li) [reply drivers/gpio/gpio_ite_it8xxx2.c]: Done.

File: drivers/gpio/gpio_ite_it8xxx2.c [hunk 6]
```diff
@@ -30,6 +30,8 @@ struct gpio_ite_cfg {
 	/* gpio port output type register (bit mapping to pin) */
 	uintptr_t reg_gpotr;
 	uint8_t ngpios;
+	/* GPIO group index */
+	uint8_t idx;
```
Comment (keith-zephyr) [inline drivers/gpio/gpio_ite_it8xxx2.c:34]: Nit - can you rename idx to index so it matches the devicetree property name?

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_ite_it8xxx2.c (hunk 2 above)
Comment (mnkp) [reply drivers/gpio/gpio_ite_it8xxx2.c:352]: The __ASSERT has to be called before the variable is used as an index.

I've made one mistake. The value of the `pin` argument is verified in `include/drivers/gpio.h` so there is no need to do it here again. Only gpio_config->idx needs to be verified (unless you're able to remove it completely as suggested by @galak).

Comment (Dino-Li) [reply drivers/gpio/gpio_ite_it8xxx2.c]: Done. (change to verify gpio_config->index only).

File: drivers/gpio/gpio_ite_it8xxx2.c [hunk 7]
```diff
@@ -221,6 +223,120 @@ static const struct {
...
+		[4] = {&IT8XXX2_GPIO_GCR24, BIT(0)},
+		[5] = {&IT8XXX2_GPIO_GCR24, BIT(1)},
+		[6] = {&IT8XXX2_GPIO_GCR24, BIT(5)},
+		[7] = {&IT8XXX2_GPIO_GCR24, BIT(6)} },
+	[GPIO_GROUP_INDEX(gpiob)] = {
+		[3] = {&IT8XXX2_GPIO_GCR22, BIT(1)},
+		[4] = {&IT8XXX2_GPIO_GCR22, BIT(0)},
+		[5] = {&IT8XXX2_GPIO_GCR19, BIT(7)},
+		[6] = {&IT8XXX2_GPIO_GCR19, BIT(6)},
+		[7] = {&IT8XXX2_GPIO_GCR24, BIT(4)} },
```
[the remark this answers is not in the document]
Comment (Dino-Li) [reply drivers/gpio/gpio_ite_it8xxx2.c:245]: Yes, the datasheet will be updated later.

File: drivers/gpio/gpio_ite_it8xxx2.c [hunk 8]
```diff
@@ -278,10 +400,17 @@ static int gpio_ite_configure(const struct device *dev,
...
+	/* 1.8V or 3.3V */
+	if (reg_1p8v != NULL) {
+		gpio_flags_t volt = flags & GPIO_VOLTAGE_MASK;
+
+		if (volt == GPIO_VOLTAGE_1P8) {
+			*reg_1p8v |= mask_1p8v;
+		} else if (volt == GPIO_VOLTAGE_3P3 ||
+			   volt == GPIO_VOLTAGE_DEFAULT) {
+			*reg_1p8v &= ~mask_1p8v;
+		}
```
[the remark this answers is not in the document]
Comment (Dino-Li) [reply drivers/gpio/gpio_ite_it8xxx2.c]: Done.

File: drivers/gpio/gpio_ite_it8xxx2.c [hunk 9]
```diff
@@ -278,10 +400,17 @@ static int gpio_ite_configure(const struct device *dev,
 				GPCR_PORT_PIN_MODE_PULLDOWN);
 	}
 
-	/*
-	 * TODO: There are some gpios are 1.8v input at default.
-	 * Is there a configuration flag for 1.8/3.3v selection.
-	 */
+	/* 1.8V or 3.3V */
```
[the remark this answers is not in the document]
Comment (Dino-Li) [reply drivers/gpio/gpio_ite_it8xxx2.c]: > Should the voltage level be set before enabling the output signal? Is there a possibility of applying a 3.3V on a pin temporarily that is connected to a device that is only 1.8V tolerant?
> 
Done. (set the voltage level before enabling the output)
> Also - the datasheet indicates that the internal pull-up can't be set to 1 if 1.8V voltage is used. Can you add a check for this?
Done.

File: drivers/gpio/gpio_ite_it8xxx2.c (hunk 6 above)
Comment (Dino-Li) [reply drivers/gpio/gpio_ite_it8xxx2.c]: Done.

File: drivers/gpio/gpio_ite_it8xxx2.c [hunk 10]
```diff
@@ -221,6 +223,120 @@ static const struct {
...
+#define GPIO_GROUP_COUNT 16
+#define GPIO_GROUP_INDEX(label) DT_PROP(DT_NODELABEL(label), index)
+
+/* general control registers for selecting 1.8V/3.3V */
+static const struct {
+	volatile uint8_t *reg_1p8v;
+	uint8_t mask_1p8v;
+} gpio_1p8v[GPIO_GROUP_COUNT][8] = {
+	[GPIO_GROUP_INDEX(gpioa)] = {
+		[4] = {&IT8XXX2_GPIO_GCR24, BIT(0)},
```
Comment (sjg20) [inline drivers/gpio/gpio_ite_it8xxx2.c:236]: Why use addresses here? Wouldn't it be better to store the index, then get the address at runtime?

File: dts/bindings/gpio/ite,it8xxx2-gpio.yaml [hunk 11]
```diff
@@ -18,6 +18,11 @@ properties:
...
+        type: int
+        required: true
+        description: GPIO group index
```
Comment (sjg20) [inline dts/bindings/gpio/ite,it8xxx2-gpio.yaml:20]: Please expand this, it needs a lot more detail. Index into what? Please be specific so people know what to set here

File: soc/riscv/riscv-ite/common/chip_chipregs.h [hunk 12]
```diff
@@ -1716,6 +1716,22 @@ struct flash_it8xxx2_regs {
 
 #define IT8XXX2_GPIO_GCR        ECREG(IT8XXX2_GPIO_BASE + 0x00)
 
+#define IT8XXX2_GPIO_GCR25      ECREG(IT8XXX2_GPIO_BASE + 0xD1)
```
Comment (sjg20) [inline soc/riscv/riscv-ite/common/chip_chipregs.h:1719]: lower-case hex

**Review by sjg20 — CHANGES_REQUESTED** [review]

**Review by keith-zephyr — APPROVED** [review]

File: soc/riscv/riscv-ite/common/chip_chipregs.h (hunk 12 above)
Comment (Dino-Li) [reply soc/riscv/riscv-ite/common/chip_chipregs.h]: Done

File: dts/bindings/gpio/ite,it8xxx2-gpio.yaml (hunk 11 above)
Comment (Dino-Li) [reply dts/bindings/gpio/ite,it8xxx2-gpio.yaml]: Done

File: drivers/gpio/gpio_ite_it8xxx2.c (hunk 10 above)
Comment (Dino-Li) [reply drivers/gpio/gpio_ite_it8xxx2.c]: Done

File: drivers/gpio/gpio_ite_it8xxx2.c (hunk 2 above)
Comment (mnkp) [reply drivers/gpio/gpio_ite_it8xxx2.c]: Thanks, however you still need to make sure that __ASSERT is called before the variable is used as an index. I.e. before the line with `gpio_1p8v[gpio_config->index][pin].offset`. If index value is invalid you're going to access out of bound array element.

File: drivers/gpio/gpio_ite_it8xxx2.c [hunk 13]
```diff
@@ -233,8 +349,15 @@ static int gpio_ite_configure(const struct device *dev,
 	volatile uint8_t *reg_gpdr = (uint8_t *)gpio_config->reg_gpdr;
 	volatile uint8_t *reg_gpcr = (uint8_t *)(gpio_config->reg_gpcr + pin);
 	volatile uint8_t *reg_gpotr = (uint8_t *)gpio_config->reg_gpotr;
+	volatile uint8_t *reg_1p8v = &IT8XXX2_GPIO_GCRX(
+				gpio_1p8v[gpio_config->index][pin].offset);
+	volatile uint8_t mask_1p8v =
+				gpio_1p8v[gpio_config->index][pin].mask_1p8v;
```
Comment (brockus-zephyr) [inline drivers/gpio/gpio_ite_it8xxx2.c:355]: declare the variables reg_1p8v/mask_1p8v here without the assignment

File: drivers/gpio/gpio_ite_it8xxx2.c [hunk 14]
```diff
@@ -249,6 +372,22 @@ static int gpio_ite_configure(const struct device *dev,
 	else
 		*reg_gpotr &= ~mask;
 
+	/* 1.8V or 3.3V */
```
Comment (brockus-zephyr) [inline drivers/gpio/gpio_ite_it8xxx2.c:375]: assign the values to reg_1p8v/mask_1p8v here.

**Review by brockus-zephyr — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_ite_it8xxx2.c (hunk 2 above)
Comment (Dino-Li) [reply drivers/gpio/gpio_ite_it8xxx2.c]: Done.

File: drivers/gpio/gpio_ite_it8xxx2.c (hunk 13 above)
Comment (Dino-Li) [reply drivers/gpio/gpio_ite_it8xxx2.c]: Done.

File: drivers/gpio/gpio_ite_it8xxx2.c (hunk 14 above)
Comment (Dino-Li) [reply drivers/gpio/gpio_ite_it8xxx2.c:369]: Done.

**Review by brockus-zephyr — APPROVED** [review]

**Review by brockus-zephyr — APPROVED** [review]
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/36443)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @galak, @GTLin08, @katsuster, @kgugala, @mbolivar-nordic, @mgielda, @mnkp, @pgielda, and @sjg20)

File: drivers/gpio/gpio_ite_it8xxx2.c [hunk 15]
```diff
@@ -440,6 +575,7 @@ static const struct gpio_ite_cfg gpio_ite_cfg_##inst = {           \
 	.reg_gpcr = DT_INST_REG_ADDR_BY_IDX(inst, 1),              \
 	.reg_gpdmr = DT_INST_REG_ADDR_BY_IDX(inst, 2),             \
 	.reg_gpotr = DT_INST_REG_ADDR_BY_IDX(inst, 3),             \
+	.index = DT_PROP(DT_INST(inst, ite_it8xxx2_gpio), index),  \
```
Comment (mnkp) [inline drivers/gpio/gpio_ite_it8xxx2.c:578]: You should use `DT_INST_PROP` here.

Comment (Dino-Li) [reply drivers/gpio/gpio_ite_it8xxx2.c]: Done.

**Comment by Dino-Li** [comment]
rebase.

**Review by mnkp — APPROVED** [review]

**Review by sjg20 — APPROVED** [review]

**Comment by brockus-zephyr** [comment]
@galak are your concerns addressed and changes requested made?

**Review by brockus-zephyr — APPROVED** [review]
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/36443)* status: all files reviewed, 1 unresolved discussion (waiting on @galak, @GTLin08, @katsuster, @kgugala, @mbolivar-nordic, @mgielda, @mnkp, and @pgielda)

File: drivers/gpio/gpio_ite_it8xxx2.c [hunk 16]
```diff
@@ -222,6 +224,120 @@ static const struct {
 };
 BUILD_ASSERT(ARRAY_SIZE(gpio_irqs) == IT8XXX2_IRQ_COUNT + 1);
 
+/* gpio group a, b, c, d, e, f, g, h, i, j, k ,l ,m, ksi, ksoh, and ksol */
+#define GPIO_GROUP_COUNT 16
+#define GPIO_GROUP_INDEX(label) DT_PROP(DT_NODELABEL(label), index)
```
Comment (brockus-zephyr) [inline drivers/gpio/gpio_ite_it8xxx2.c:229]: Change this to the following:
```C
#define GPIO_GROUP_INDEX(label) \
                (uint8_t)(DT_REG_ADDR(DT_NODELABEL(label)) - \
                          DT_REG_ADDR(DT_NODELABEL(gpioa)))
```

File: drivers/gpio/gpio_ite_it8xxx2.c [hunk 17]
```diff
@@ -440,6 +575,7 @@ static const struct gpio_ite_cfg gpio_ite_cfg_##inst = {           \
 	.reg_gpcr = DT_INST_REG_ADDR_BY_IDX(inst, 1),              \
 	.reg_gpdmr = DT_INST_REG_ADDR_BY_IDX(inst, 2),             \
 	.reg_gpotr = DT_INST_REG_ADDR_BY_IDX(inst, 3),             \
+	.index = DT_INST_PROP(inst, index),                        \
```
Comment (brockus-zephyr) [inline drivers/gpio/gpio_ite_it8xxx2.c:578]: Change this to the following:
~~~C
        .index = (uint8_t)(DT_INST_REG_ADDR(inst) -                \
                           DT_REG_ADDR(DT_NODELABEL(gpioa))),      \
~~~

File: dts/bindings/gpio/ite,it8xxx2-gpio.yaml (hunk 5 above)
Comment (brockus-zephyr) [inline dts/bindings/gpio/ite,it8xxx2-gpio.yaml:17]: Remove 17-21. Index will no longer be in the DeviceTree

File: dts/riscv/it8xxx2.dtsi [hunk 18]
```diff
@@ -292,6 +292,7 @@
 			ngpios = <8>;
 			label = "GPIO_A";
 			gpio-controller;
+			index = <0x0>;
```
Comment (brockus-zephyr) [inline dts/riscv/it8xxx2.dtsi:295]: remove all of the
~~~C
                index=<X>
~~~

**Review by brockus-zephyr — CHANGES_REQUESTED** [review]
I believe making these changes will address the requests from galak@ and should be ready to land.

**Comment by Dino-Li** [comment]
> the dt index property doesn't seen required as seems like `0xf01601 - reg` will get you this.

Done.

File: drivers/gpio/gpio_ite_it8xxx2.c (hunk 16 above)
Comment (Dino-Li) [reply drivers/gpio/gpio_ite_it8xxx2.c]: Done.

File: drivers/gpio/gpio_ite_it8xxx2.c (hunk 17 above)
Comment (Dino-Li) [reply drivers/gpio/gpio_ite_it8xxx2.c]: Done.

File: dts/bindings/gpio/ite,it8xxx2-gpio.yaml (hunk 5 above)
Comment (Dino-Li) [reply dts/bindings/gpio/ite,it8xxx2-gpio.yaml]: Done.

File: dts/riscv/it8xxx2.dtsi (hunk 18 above)
Comment (Dino-Li) [reply dts/riscv/it8xxx2.dtsi]: Done.

**Review by brockus-zephyr — APPROVED** [review]
@galak I believe the most recent update removes the index from the tree as you requested.

**Review by brockus-zephyr — COMMENTED** [review]
> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/36443)* status: all files reviewed, 5 unresolved discussions (waiting on @brockus-zephyr, @galak, @GTLin08, @katsuster, @kgugala, @mbolivar-nordic, @mgielda, @mnkp, and @pgielda)

**Comment by brockus-zephyr** [comment]
Can this be merged? The only outstanding issue was raised by @galak and that has long since been resolved.

**Review by brockus-zephyr — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/36443#-:-MeaG_fiBmh8nvVaWz4I:bnfp4nl)* status: all files reviewed, 5 unresolved discussions (waiting on @brockus-zephyr, @galak, @GTLin08, @katsuster, @kgugala, @mbolivar-nordic, @mgielda, @mnkp, and @pgielda)
